### PR TITLE
From q refactor

### PR DESCRIPTION
**Pull request by MoogSven**
**Description**
The `AsInteger` trait is intended to be used in other parts of the code as well.
It abstracts rust integers, `Z` and `Zq` into something that can be converted to `fmpz`.

Intended pattern if just a `&fmpz` is required (does not compile):
```rust
match a.get_fmpz_ref(){
    None => let v = a.into_fmpz(); [...]; fmpz_clear(v),
    v => [...]
}
```
Maybe a better syntax is possible with closures?

This PR ...
- [x] Adds `AsInteger` trait.
    - [x] Implements `AsInteger` for rust integers
    - [x] Implements `AsInteger` for `Z`
    - [x] Implements `AsInteger` for `Zq`
- [x] Reworks `Q::from((a,b))` to accept all combinations of borrowed/owned and the different Integer types.

**Testing**
- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

**Comment by jnsiemer** [comment]
Rebase to accomplish a passing pipeline

File: src/rational/q/from.rs
Comment (Marvin-Beckmann) [inline src/rational/q/from.rs:401]: Generally like extensive testing, but it looks a little too overklill, but I dont mind too much, lave it imo

File: src/rational/q/from.rs
Comment (Marvin-Beckmann) [inline src/rational/q/from.rs:603]: I would say yes and leave it. With more personal consideration, this can remain a panic imo

**Review by Marvin-Beckmann — CHANGES_REQUESTED** [review]
We could consider adding an explanation for the AsInteger trait on the first level, i.e. first module level, since it does not occur in the documentation and would allows users to get a glimpse of what this means.

**Review by Marvin-Beckmann — COMMENTED** [review]
@jnsiemer can you have a look over the dedicated memory tests, you are probably better equipped to review them

File: src/integer/z/fmpz_helpers.rs
Comment (jnsiemer) [inline src/integer/z/fmpz_helpers.rs:220]: If you want to check whether the memory is overwritten, check for `assert_eq(_a.value.0, z.value.0)`. For Z this should be fairly convenient. Just carry the i64 `z.value.0` out of the inner block if needed.
If you want to check whether the new value isn't stored at the same point in memory use `assert_ne` with the new value from `from_fmpz` for its `value.0`.

File: src/integer/z/fmpz_helpers.rs
Comment (jnsiemer) [inline src/integer/z/fmpz_helpers.rs:210]: I don't understand why this shouldn't work for owned types as the `z` is dropped afterward anyways.
```suggestion
    /// This can not be tested for an owned [`Z`], since that would violate the
    /// ownership constraint -> not compiling.
```

**Review by jnsiemer — CHANGES_REQUESTED** [review]
I love the new generic solution for instantiations of `Q`.
The new functions introduced via `AsInteger` introduce the need for the developer to actively think about memory issues (as also stated in the trait doc comment). The functions are `unsafe`, which is a good indicator that we should think about this stuff when using it. As far as I can tell, there should be no memory issues in the current use of these functions, which does not mean that they can be introduced in the future.
There are a few additional questions / suggestions regarding the fmpz_helpers, which should probably be thought about in the fmpz_mod_helpers file, too.
Great work! I like the approach :)

File: src/integer_mod_q/z_q/fmpz_mod_helpers.rs
Comment (MoogSven) [reply src/integer_mod_q/z_q/fmpz_mod_helpers.rs:54]: Like Niklas said, it just calls the code for `Z`. Therefore, I will leave it this way.
It could be changed to
```rust
(&self.value).into_fmpz()
``` 
But I am not sure if that looks better.

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: You are right. That was still from an old implementation where I compared to the original.
I also added tests for references now.

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: That is a much cleaner way to test it.
I reworked the tests to just check the i64 representation of fmpz now.

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: Added, also for the `u64` case.

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: removed empty line

File: src/integer_mod_q/z_q/fmpz_mod_helpers.rs
Comment (MoogSven) [reply src/integer_mod_q/z_q/fmpz_mod_helpers.rs]: done

File: src/rational/q/from.rs
Comment (MoogSven) [reply src/rational/q/from.rs]: ok, I removed the comment

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: snake case should be used for macro names.
Renamed it to `implement_as_integer_over_i64` (maybe a bit long, but also just used once). Also added a short doc comment describing the macro

File: src/integer_mod_q/z_q/fmpz_mod_helpers.rs
Comment (MoogSven) [reply src/integer_mod_q/z_q/fmpz_mod_helpers.rs]: Added, also changed fmpz to [`fmpz`]

File: src/integer/z/fmpz_helpers.rs
Comment (MoogSven) [reply src/integer/z/fmpz_helpers.rs]: Added, also changed fmpz to [`fmpz`]

**Comment by MoogSven** [comment]
> We could consider adding an explanation for the AsInteger trait on the first level, i.e. first module level, since it does not occur in the documentation and would allows users to get a glimpse of what this means.

Sounds like an interesting idea. We probably need a module description for lib.rs anyway. Added a [kanboard item](https://github.com/orgs/qfall/projects/1/views/1?pane=issue&itemId=28155274) for this.

File: src/integer_mod_q/z_q/fmpz_mod_helpers.rs
Comment (Marvin-Beckmann) [reply src/integer_mod_q/z_q/fmpz_mod_helpers.rs]: From what I understand, here you take the reference, i.e. not the efficient implementation, where you swap the value, compared to what you do for the Z value. But I am fine with this too, but this would not be the efficient variant (at least I think so :D)

**Review by Marvin-Beckmann — APPROVED** [review]
Looks good from my side. Please also wait for Niklas approval

File: src/integer_mod_q/z_q/fmpz_mod_helpers.rs
Comment (MoogSven) [reply src/integer_mod_q/z_q/fmpz_mod_helpers.rs]: I understand what you mean now. I removed the `&` and it is now using the optimized implementation when ownership is transferred.

**Review by jnsiemer — APPROVED** [review]

**Pull request merged**
